app/objects/c_agent.py

```diff
@@ -13,47 +13,58 @@
-    host = ma.fields.String()
     watchdog = ma.fields.Integer()
-    contact = ma.fields.String()
-    pending_contact = ma.fields.String()
-    links = ma.fields.List(ma.fields.Nested(LinkSchema()))
-    proxy_receivers = ma.fields.Dict(keys=ma.fields.String(), values=ma.fields.List(ma.fields.String()))
-    proxy_chain = ma.fields.List(ma.fields.List(ma.fields.String()))
-    origin_link_id = ma.fields.Integer()
-    deadman_enabled = ma.fields.Boolean()
-    available_contacts = ma.fields.List(ma.fields.String())
-    created = ma.fields.DateTime(format='%Y-%m-%d %H:%M:%S')
-    host_ip_addrs = ma.fields.List(ma.fields.String())
+    group = ma.fields.String(required=False)
+    architecture = ma.fields.String(required=False)
+    platform = ma.fields.String(required=False)
+    server = ma.fields.String(required=False)
+    upstream_dest = ma.fields.String(required=False, allow_none=True)
+    username = ma.fields.String(required=False)
+    location = ma.fields.String(required=False)
+    pid = ma.fields.Integer(required=False)
+    ppid = ma.fields.Integer(required=False)
+    trusted = ma.fields.Boolean(required=False)
+    executors = ma.fields.List(ma.fields.String(), required=False)
+    privilege = ma.fields.String(required=False)
+    exe_name = ma.fields.String(required=False)
+    host = ma.fields.String(required=False)
+    contact = ma.fields.String(required=False)
+    proxy_receivers = ma.fields.Dict(keys=ma.fields.String(), values=ma.fields.List(ma.fields.String()),
+                                     required=False, allow_none=True)
+    proxy_chain = ma.fields.List(ma.fields.List(ma.fields.String()), required=False, allow_none=True)
+    origin_link_id = ma.fields.Integer(required=False)
+    deadman_enabled = ma.fields.Boolean(required=False, allow_none=True)
+    available_contacts = ma.fields.List(ma.fields.String(), required=False, allow_none=True)
+    host_ip_addrs = ma.fields.List(ma.fields.String(), required=False, allow_none=True)
+
+    display_name = ma.fields.String(dump_only=True)
+    created = ma.fields.DateTime(format='%Y-%m-%d %H:%M:%S', dump_only=True)
+    last_seen = ma.fields.DateTime(format='%Y-%m-%d %H:%M:%S', dump_only=True)
+    links = ma.fields.List(ma.fields.Nested(LinkSchema), dump_only=True)
+    pending_contact = ma.fields.String(dump_only=True)
 
     @ma.pre_load
     def remove_nulls(self, in_data, **_):
         return {k: v for k, v in in_data.items() if v is not None}
 
+    @ma.pre_load
+    def remove_properties(self, data, **_):
+        data.pop('display_name', None)
+        data.pop('created', None)
+        data.pop('last_seen', None)
+        data.pop('links', None)
+        data.pop('pending_contact', None)
+        return data
+
 
 class AgentSchema(AgentFieldsSchema):
 
     @ma.post_load
-    def build_agent(self, data, **_):
-        return Agent(**data)
+    def build_agent(self, data, **kwargs):
+        return None if kwargs.get('partial') is True else Agent(**data)
```

Comment:
Is there a reason we explicitly validate against true?

Reply:
Forgetting exactly how, but at some point I thought I had seen both `False` and `None` values for `partial`. Figured an explicit check would be better in case arbitrary values could be set.

Reply:
Makes sense, though for python `None`, `False`, `[]`, `{}`, `''`, `0`, and more will all evaluate to false in the case of truth value testing.  So depending on what might be coming back in partial it may be covered.

https://docs.python.org/3.7/library/stdtypes.html#truth-value-testing

Reply:
Yeah, wasn't sure if arbitrary values like `'string'` or `1` could make their way here. Another values that evaluates to `True` would prevent the object from being created.